Automated Backup and Restoration System

jsmatias opened the pull request:
This backup system is designed to run within a Docker container using Cron jobs to schedule regular executions. 

**In summary, it includes:**

- **Dockerized Backup Solution**: Encapsulates the backup logic within a Docker container.

- **Cron Job Scheduling**: Utilises cron for scheduling backups at regular intervals. This approach provides flexibility in how frequently backups are performed, and for which service (MySQL, Keycloak, connectors, etc).

- **Incremental Backups**: It uses incremental backup logic implemented in cycles. Each cycle contains a full backup followed by incremental ones. Currently, it is set to run daily in weekly cycles.   

- **Restoring procedure**: The _restore.sh_ script is available to be run manually if any file needs to be restored.

jsmatias committed:
backup configuration created

jsmatias committed:
added gitkeep

jsmatias committed:
renamed script

jsmatias committed:
added logic to incremental backup

jsmatias committed:
added restore.sh script

jsmatias committed:
added paths to .env and clean up scripts

jsmatias committed:
added paths to .env and clean up scripts

jsmatias committed:
replaced some types in the xml, so the editor stops complaining

jsmatias committed:
refactored scripts and added some tests

jsmatias committed:
concluded tests plus minor changes in the scripts

jsmatias commented on src/tests/resources/connectors/zenodo/list_records_2.xml:
These changes were made just to make vscode to stop complaining about not listed `resourceTypeGeneral`.

PGijsbers commented on backups/backup.sh:
```diff
@@ -0,0 +1,73 @@
+#!/bin/bash
+
+if [ $# -ne 3 ]; then
+  echo "Usage: $0 path/to/data path/to/backup/dir <cycle length:int>"
```
I think it would be good to give a little more documentation about cycle length.

PGijsbers replied:
and a link to e.g., https://www.gnu.org/software/tar/manual/html_section/Incremental-Dumps.html

jsmatias replied:
Sure, thanks. 
I added detailed information about it for both scripts, _backup.sh_ and _restore.sh_.

jsmatias replied:
I also added a README.md to describe the process

PGijsbers commented on src/tests/backups/test_backup_scripts.py:
```diff
@@ -0,0 +1,120 @@
+        backup_cycle_1_path = backup_dir / "original" / "original_1"
+        files_in_cycle_0 = [
+            backup_cycle_0_path / "original0.tar.gz",
+            backup_cycle_0_path / "original1.tar.gz",
```
I assume this would be a file of insignificant size as there is no delta between the first and second backup?

jsmatias replied:
Yes, you're correct. If no change was made to the original files after the backup of level 0 (_original0.tar.gz_), then the next level of backup (_original1.tar.gz_) will be empty. This is information of the state of the last backup is stored in the _.snar_ files.

PGijsbers commented on src/tests/backups/test_backup_scripts.py:
```diff
@@ -0,0 +1,120 @@
+        files_in_cycle_0 = [
+            backup_cycle_0_path / "original0.tar.gz",
+            backup_cycle_0_path / "original1.tar.gz",
+            backup_cycle_0_path / "original.snar",
```
It looks like you can delete this file after completing a full cycle? Though keeping it around probably doesn't hurt as it should only be meta-data. Keeping it also reduces the complexity of the script and potentially allows for continued backups at a higher level.

jsmatias replied:
Yes, you're correct. For now, I believe it's fine to keep old backups just in case, since we're not yet assessing or validating the metadata quality. In future, it will be better to implement some sort of deletion of older backups, then we would keep just one or a few ones.

PGijsbers reviewed (changes requested):
Some minor suggestions. Mainly regarding documentation of the cycle functionality. The implementation seems OK. Did not try to run this yet.

jsmatias commented on backups/entry.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+mkdir -p /opt/backups/data/
+
+# Run cron on the foreground with log level WARN
+/usr/sbin/cron -f -l 4
```
That sounds good, I'll do it.

jsmatias committed:
moved commands from entry.sh file to docker related ones

jsmatias commented on backups/backup.sh:
```diff
@@ -0,0 +1,73 @@
+tar -czf "$backup_file" --listed-incremental="$snapshot_file" "$path_to_backup"
+
+echo $(date -u) "$data_to_backup done!"
+
```
I agree. It just got longer than I anticipated. In principle it should be just a simple `tar` command plus a few conditions. When I realised it was already around 100 lines. I'll pay better attention for future ones.

jsmatias commented on src/tests/backups/test_backup_scripts.py:
```diff
@@ -0,0 +1,120 @@
+        call_backup_command(backup_command)
+
+        call_restore_command(restore_command)
+        assert not os.path.exists(restored_file1_path), f"{restored_file1_path} shouldn't be here."
```
Well noticed! Thanks, I replaced it.

jsmatias committed:
replaced  by

jsmatias committed:
improved documentation

jsmatias committed:
Update pytest-tests.yml

Pytest is passing when it runs locally and within pre-commit, but fails when it runs again after pre-commit. I changed the order to check if it still has same behaviour.

jsmatias committed:
added conditional skip to this unit test, so it runs only if gnu tar is installed.

jsmatias committed:
Merge branch 'backup-routine' of https://github.com/aiondemand/AIOD-rest-api into backup-routine

jsmatias committed:
restored pytest workflow

jsmatias committed:
minor changed in tests

jsmatias committed:
roll back tests as they were working before

jsmatias committed:
renamed variables and minor changes

jsmatias committed:
moved restore logic to be executed from inside the container

jsmatias committed:
minor changes to restore.sh and moved tests to be run from inside the container an initial check

jsmatias committed:
improved backup documentation

jsmatias commented:
After addressing most of the feedback from previous comments, I encountered an issue where the unit tests for this PR ceased to function correctly, even when using the same commit tag that had previously passed. Furthermore, using a `subprocess` in Python to execute the `tar` command for pytests proved to be suboptimal. This approach introduced a dependency on the operating system hosting the tests, specifically due to the requirement of having GNU tar for this backup service to function correctly.

Considering the critical role of these tests, I've relocated their execution to occur inside the backup container itself, initiated from the entry point. This adjustment ensures that the backup container will only start backup operations once it has successfully passed these tests.

To circumvent the dependency on the host OS having GNU tar installed for the restoration script, I've added an `_entry_restore.sh` script. This script is designed to be executed within the container whenever a backup restoration is needed.

In summary, the following modifications have been implemented:
- Unit tests have been moved to execute from within the container, initiated at the entry point.
- The logic for data restoration has been integrated into the container environment.
- A `readme.md` file has been added to enhance the documentation, providing clear instructions on how to perform backups and restorations.

jsmatias committed:
backup and restore logic simplified and removed from container

jsmatias committed:
check path to data in backup, paths to data put in the env, and added mysql restore logic

jsmatias committed:
Added some documentation

PGijsbers commented on scripts/mysql_restore.sh:
```diff
@@ -0,0 +1,17 @@
+docker cp "${DATA_PATH}/mysql_dump/backup.sql" sqlserver:/tmp/backup.sql
+
+echo $(date -u) "Restoring database..."
+docker exec -i /bin/bash -c "mysql -uroot -p${MYSQL_ROOT_PASSWORD} aiod < /tmp/backup.sql"
```
```suggestion
docker exec -i sqlserver /bin/bash -c "mysql -uroot -p${MYSQL_ROOT_PASSWORD} aiod < /tmp/backup.sql"
```

jsmatias replied:
Applied the change in commit a2a6741.

PGijsbers reviewed (changes requested):
Hi,

I left some feedback. Overall it looks good. Unfortunately, I couldn't test this that well, it seems that `tar` on macos does not do incremental backups :/ And this does worry me, since it seems that you didn't run this either? Because a critical part of the `mysql_restore` call is missing (the docker container name).

I can test this in a linux environment later, but figured I will send this feedback already.

jsmatias commented:
> Hi,
> 
> I left some feedback. Overall it looks good. Unfortunately, I couldn't test this that well, it seems that `tar` on macos does not do incremental backups :/ And this does worry me, since it seems that you didn't run this either? Because a critical part of the `mysql_restore` call is missing (the docker container name).
> 
> I can test this in a linux environment later, but figured I will send this feedback already.

Indeed, `tar` on MacOS doesn't have incremental backups. But I installed `GNU tar` to to run it locally. 
In fact, the docker container name was missing! Thanks for spotting that. 
What happened is that when I was testing it, I ran the command just once directly from terminal and when I passed the commands to the script I forgot the container's name. I ended up not testing it again, because the process to restore the entire DB takes some time. 

I cleaned up the DB to make it quicker and tested it properly this time.

jsmatias committed:
Update scripts/mysql_restore.sh

jsmatias committed:
Update scripts/README.md

Co-authored-by: Pieter Gijsbers <p.gijsbers@tue.nl>

jsmatias commented on scripts/README.md:
```diff
@@ -0,0 +1,76 @@
+
+### `mysql_dump.sh`
+
+- **Purpose**: Exports a logical backup of the `aiod` database from the SQL server and extracts it to `data/path/mysql_dump`.
```
Did you mean to add some change here?

PGijsbers replied:
I am not sure (anymore) what happened here. Please ignore - likely an artifact from me restructuring some suggestions half-way through the document.

jsmatias committed:
Update scripts/README.md

Co-authored-by: Pieter Gijsbers <p.gijsbers@tue.nl>

jsmatias committed:
Update scripts/README.md

Co-authored-by: Pieter Gijsbers <p.gijsbers@tue.nl>

jsmatias commented on scripts/README.md:
```diff
@@ -0,0 +1,76 @@
+- **Usage**: Simply run the script without any arguments.
+
+
+### `database-connect.sql`
```
Good point. I updated the file name as well.

jsmatias committed:
Update scripts/README.md

Co-authored-by: Pieter Gijsbers <p.gijsbers@tue.nl>

jsmatias committed:
rename database-connect script .sql -> .sh

jsmatias commented on scripts/README.md:
```diff
@@ -0,0 +1,76 @@
+
+### `clean.sh`
+
+**Warning!** - This deletes all the content in the data/ directory!
```
Cool! I added it.

jsmatias commented on scripts/realm_export.sh:
```diff
@@ -0,0 +1,18 @@
+    mkdir "$LOCAL_BACKUP_PATH"
+fi
+
+docker cp keycloak:/tmp/aiod.json "$LOCAL_BACKUP_PATH"/aiod.json
```
Sure, we can do that. 👍

jsmatias committed:
Update scripts/realm_export.sh

Co-authored-by: Pieter Gijsbers <p.gijsbers@tue.nl>

jsmatias committed:
Update scripts/README.md

jsmatias commented on scripts/mysql_dump.sh:
```diff
@@ -0,0 +1,22 @@
+"
+
+if [ ! -d "$LOCAL_BACKUP_PATH" ]; then
+    mkdir "$LOCAL_BACKUP_PATH"
```
It's fine. This path should be dedicated to the backups only. If a file with same name is there, this should be addressed before proceeding with the backups. This way any conflict can be avoided.

jsmatias commented on scripts/realm_export.sh:
```diff
@@ -0,0 +1,18 @@
+docker exec -i keycloak /bin/bash -c "/opt/keycloak/bin/kc.sh export --file /tmp/aiod.json --realm aiod"
+
+if [ ! -d "$LOCAL_BACKUP_PATH" ]; then
+    mkdir "$LOCAL_BACKUP_PATH"
```
It's okay, same as above.

jsmatias committed:
Update scripts/restore.sh

Co-authored-by: Pieter Gijsbers <p.gijsbers@tue.nl>

jsmatias committed:
simplified case statement to if statement in the restore.sh

jsmatias commented on scripts/restore.sh:
```diff
@@ -0,0 +1,111 @@
+    echo "   Restoring: ${backup_file}"
+    tar --directory="$destination_dir" --extract --file="$backup_file_path" --listed-incremental=/dev/null
+done
+;;
```
Indeed. This was a bit messy, I moved it out and replaced it by an `if` statement in the commit [18449b3](https://github.com/aiondemand/AIOD-rest-api/pull/275/commits/18449b3b65707d0b21abd5637f0cce5fe5feb813).

jsmatias committed:
add skipif to tests in case GNU tar isn't installed in the system

jsmatias commented on src/tests/backups/test_backup_restore.py:
```diff
@@ -0,0 +1,119 @@
+    return subprocess.run(command, env=env, input="y\n", text=True, check=True)
+
+
+def test_backup_happy_path():
```
Yes, the scripts require GNU tar. I added a condition to check if GNU tar is installed and available in the system, otherwise the tests are skipped.

jsmatias committed:
Added comments to restore.sh

jsmatias commented on scripts/restore.sh:
```diff
@@ -0,0 +1,111 @@
+        level=$((i - 1))
+    fi
+    i=$((i + 1))
+done
```
I agree, this is not very intuitive. I added some comments to the file in commit [3b144ba](https://github.com/aiondemand/AIOD-rest-api/pull/275/commits/3b144bae6543afe1a8c1a80c893c344e9871ec12).

jsmatias committed:
fixed lexigraphical sort

PGijsbers commented on scripts/backup.sh:
```diff
@@ -0,0 +1,75 @@
+        level=0
+    fi
+fi
+
```
Printing some information on which cycle and level are used will be useful information to have in the logs.

PGijsbers commented on scripts/restore.sh:
```diff
@@ -0,0 +1,109 @@
+check_file_or_dir $backup_dir
+check_file_or_dir $destination_dir
+
+echo ""
+echo "Are you sure you want to proceed? (y/n)"
+echo "This overwrites all files in the destination directory and deletes the ones which are not in the archive!"
+echo ""
+echo "$backup_dir -------> $destination_dir/$data_to_restore"
```
```suggestion
echo ""
echo "$backup_dir -------> $destination_dir/$data_to_restore"
echo ""
echo "This overwrites all files in the destination directory and deletes the ones which are not in the archive!"
echo "Are you sure you want to proceed? (y/n)"
```

The prompt is easier to miss if it's not the last thing on the screen.

jsmatias replied:
Nice! Good point.

PGijsbers reviewed (approved):
Got around to testing it (sorry for the delay). Seems to work as expected/described (tested only with the database backup, adding/removing rows). Left minor comments from actually using it, but feel free to merge without a new review round.

jsmatias committed:
Update scripts/README.md

Co-authored-by: Pieter Gijsbers <p.gijsbers@tue.nl>

jsmatias committed:
Update scripts/restore.sh

Co-authored-by: Pieter Gijsbers <p.gijsbers@tue.nl>

jsmatias committed:
add output message about cycle and level, and tar version check before the scripts start.
